Allow setting specific address where for service to bind to

rmescandon opened the pull request:
snap.etcd.etcd.service can now listen only in unit private or public address, or in any, as so far.

config: added service_network_interface parameter
reactive: react to service_network_interface config changed
templates.etcd2 and templates.etcd3: updated listening addresses

rmescandon committed:
Allow setting specific address where for service to bind to

config: added service_network_interface parameter
reactive: react to service_network_interface config changed
templates.etcd2 and templates.etcd3: updated listening addresses

hyperbolic2346 commented on config.yaml:
```diff
@@ -20,3 +20,11 @@ options:
+    description: |
+      The interface the service binds to. Valid values are 'public' or 'private'
+      to bind to public or private unit address respectively.
+      If an empty or an invalid value is set, the service binds to all network
```
Are we sure that invalid values binding to all network interfaces is the right call here?

hyperbolic2346 commented on lib/etcd_databag.py:
```diff
@@ -94,3 +98,16 @@ def storage_path(self):
+        private address is considered. Otherwise, service is bound to any
+        available network interface'''
+        service_network_iface = config('service_network_interface')
+        if 'private' in service_network_iface:
```
This will be true if I set it to 'totallybrokenbuthasprivateinsideit'. I would think we would just want to compare the string to private and public.

hyperbolic2346 commented on reactive/etcd.py:
```diff
@@ -177,6 +177,17 @@ def follower_config_changed():
 
+@when('snap.installed.etcd')
+@when('config.changed.service_network_interface')
+@when_not('etcd.installed')
```
can't change the config after etcd.installed is set?

Cynerva replied:
The `etcd.installed` state is actually an old carryover from pre-snap deployments. Today's deployments do not have it set.

A lot of handlers have `@when_not('etcd.installed')` slapped on them as a way to prevent any handlers from running until the user has migrated to snaps.

There's a note about it here:

https://github.com/juju-solutions/layer-etcd/blob/7307aa988db33687aa8707ae646d833f6bb80886/reactive/etcd.py#L45-L49

hyperbolic2346 replied:
now on revision 126 ><

rmescandon committed:
Create 'bind_to_all_interfaces' flag to bind
the service to all available network
interfaces or only the first found
bind address
